### PR TITLE
Refactor WebexBotServer to Mindmeld

### PR DESCRIPTION
**Pull request by ritvikshrivastava**
A sample server for Webex Teams integration

### CODE REVIEW

File: mindmeld/webex_bot_server.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import logging
+import json
+from flask import request
+import requests
+from ciscosparkapi import CiscoSparkAPI
+
+
+class WebexBotServer:
+
+    def __init__(self, app, WEBHOOK_ID, ACCESS_TOKEN, conv):
```
Comment (minhtuev) [inline mindmeld/webex_bot_server.py]: by convention, we should use snake case for params (webhook_id, access_token)

File: mindmeld/webex_bot_server.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        self.app = app
+        self.WEBHOOK_ID = WEBHOOK_ID
```
Comment (minhtuev) [inline mindmeld/webex_bot_server.py]: Same for class vars

File: mindmeld/webex_bot_server.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+"""
+This module contains the Webex Bot Server component.
+"""
+
+import logging
+import json
+from flask import request
+import requests
+from ciscosparkapi import CiscoSparkAPI
```
Comment (minhtuev) [inline mindmeld/webex_bot_server.py]: We should add this to our dependency

File: mindmeld/webex_bot_server.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class WebexBotServer:
```
Comment (minhtuev) [inline mindmeld/webex_bot_server.py]: Add some documentation here

**Comment by minhtuev** [comment]
Thanks Ritvik. I added some comments, mostly style changes and linting issues.

We should move it to its own subdirectory: `bot/webex_bot_server.py` so that we can import like this `from mindmeld.bot import WebexBotServer`.

Can you add a documentation for MindMeld as well? Check under `/sources`.

File: mindmeld/bot/webex_bot_server.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    def __init__(self, app, webhook_id, access_token, conv):
+        self.app = app
+        self.webhook_id = webhook_id
+        self.access_token = access_token
+        self.conv = conv
+
+        self.logger = logging.getLogger(__name__)
+
+        if not self.webhook_id or not self.access_token:
+            raise Exception('webhook_id and bot_access_token are not set')
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: Have an exception for each parameter not being set.

File: mindmeld/bot/webex_bot_server.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+        self.spark_api = CiscoSparkAPI(self.access_token)
+        self.ACCESS_TOKEN_WITH_BEARER = 'Bearer ' + self.access_token
+        self.CISCO_API_URL = 'https://api.ciscospark.com/v1'
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: These should be global constants that can be formatted based on the access_token

File: mindmeld/bot/webex_bot_server.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            # of answering it's own questions.
+            if person_id != me.id:
+                self._post_message(room_id, self.conv.say(message)[0])
+            return 'OK'
+
+    def run(self, host='localhost', port=7150):
+        self.app.run(host=host, port=port)
+
+    def _url(self, path):
+        return self.CISCO_API_URL + path
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: Please use `.format` for this and make CISCO_API_URL a global variable.

File: mindmeld/bot/webex_bot_server.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+    def _get_message(self, msg_id):
+        headers = {'Authorization': self.ACCESS_TOKEN_WITH_BEARER}
+        resp = requests.get(self._url('/messages/{0}'.format(msg_id)), headers=headers)
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: Use `.format` here as well

File: mindmeld/bot/webex_bot_server.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        self.CISCO_API_URL = 'https://api.ciscospark.com/v1'
+
+        @self.app.route('/', methods=['POST'])
+        def handle_message(self):
+            me = self.spark_api.people.me()
+            data = request.get_json()
+
+            for key in ['personId', 'id', 'roomId']:
+                if key not in data['data'].keys():
+                    return 'OK'
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: Change this to a 400s http code and issue a better error message. Moreover, these three lines can be condensed to one line like this:
```
if len(data['data'].keys() & set(['personId', 'id', 'roomId'])) == len(['personId', 'id', 'roomId']):
    return 400, payload
```

File: mindmeld/bot/webex_bot_server.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+            if data['id'] != self.webhook_id:
+                self.logger.debug("Retrieved webhook_id {} doesn't match".format(data['id']))
+                return 'OK'
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: Please change the response code and payload

File: mindmeld/bot/webex_bot_server.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+            person_id = data['data']['personId']
+            msg_id = data['data']['id']
+            txt = self._get_message(msg_id)
+            room_id = data['data']['roomId']
+
+            if 'text' not in txt:
+                return 'OK'
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: Please change the response code and payload

File: mindmeld/bot/webex_bot_server.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+            message = str(txt['text']).lower()
+
+            # Ignore the bot's own responses, else it would go into an infinite loop
+            # of answering it's own questions.
+            if person_id != me.id:
+                self._post_message(room_id, self.conv.say(message)[0])
+            return 'OK'
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: remove this line

Comment (ritvikshrivastava) [reply mindmeld/bot/webex_bot_server.py]: Throws a 500 error if this line is removed. There's an expected return from the function I believe.

File: mindmeld/bot/webex_bot_server.py (hunk 9 above)
Comment (ritvikshrivastava) [reply mindmeld/bot/webex_bot_server.py]: This condensation does not have the intended effect (logically it should though). So I have kept it the same for now.

File: setup.py [hunk 13]
```diff
@@ -34,7 +34,8 @@
     'sklearn-crfsuite>=0.3.6,<1.0',
     'immutables~=0.9',
     'tensorflow~=1.2; python_version < "3.7"',
-    'tensorflow>=1.13.1,<2.0; python_version >= "3.7"'
+    'tensorflow>=1.13.1,<2.0; python_version >= "3.7"',
+    'ciscosparkapi'
```
Comment (minhtuev) [inline setup.py]: You should make ciscosparkapi an optional dependency:

```
 extras_require={
        'bot': ['ciscosparkapi'],
    },
```
https://github.com/psf/requests/blob/3c1d36b827417fdeaf5a1c106129de30dac371d7/setup.py#L98

Comment (minhtuev) [reply setup.py]: Try to compile the library and verify that the following syntax can work:
```
pip install mindmeld[bot]
```

Comment (ritvikshrivastava) [reply setup.py]: Tested it out using `pip install -e .[bot]`
It's working fine.

File: mindmeld/bot/webex_bot_server.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+            if 'text' not in txt:
+                return 'BAD REQUEST', 400, {'message': 'Query not found'}
+
+            message = str(txt['text']).lower()
+
+            # Ignore the bot's own responses, else it would go into an infinite loop
+            # of answering it's own questions.
+            if person_id == me.id:
+                return 'OK', 200, {'message': 'Query replicating bot response'}
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: Its probably clearer to say `Input query is the bot's previous message, so don't send it to the bot again`

File: mindmeld/bot/webex_bot_server.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        self.access_token_with_bearer = ACCESS_TOKEN_WITH_BEARER + self.access_token
+
+        @self.app.route('/', methods=['POST'])
+        def handle_message():
+            me = self.spark_api.people.me()
+            data = request.get_json()
+
+            for key in ['personId', 'id', 'roomId']:
+                if key not in data['data'].keys():
+                    return 'BAD REQUEST', 400, {'message': 'personId/id/roomID key not found'}
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: Convert such messages into the following:
`BAD_REQUEST_NAME = 'BAD REQUEST'`
`BAD_REQUEST_CODE = 400`

and then utilize them.

**Comment by minhtuev** [comment]
Remember to do squash-merge

File: mindmeld/bot/webex_bot_server.py [hunk 16]
```diff
@@ -42,26 +44,31 @@ def handle_message():
...
+                payload = {'message': 'Query not found'}
+                return BAD_REQUEST_NAME, BAD_REQUEST_CODE, payload
 
             # Ignore the bot's own responses, else it would go into an infinite loop
             # of answering it's own questions.
             if person_id == me.id:
-                return 'OK', 200, {'message': 'Query replicating bot response'}
+                payload = {'message': "Input query is the bot's previous message, \
+                            so don't send it to the bot again"}
+                return 'OK', 200, payload
```
Comment (vijay120) [inline mindmeld/bot/webex_bot_server.py]: Please do the same for code 200

**Pull request merged**